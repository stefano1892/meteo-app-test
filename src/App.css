.App {
  font-family: "Mulish", sans-serif;
  background-color: #1E202B;
  color: white;
}

@import url(./styles/index.css);