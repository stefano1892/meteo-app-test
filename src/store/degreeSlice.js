import { createSlice } from "@reduxjs/toolkit";

const initialState = {
    degree: "Celsius"
}

export const degreeSlice = createSlice({
    name: 'degreeSelection',
    initialState,
    reducers: {
        changeDegree: (state, action) => {
            state.degree = action.payload
        }
    }
})

export const { changeDegree }  = degreeSlice.actions;

export default degreeSlice.reducer;