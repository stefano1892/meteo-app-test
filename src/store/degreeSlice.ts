import { createSlice } from "@reduxjs/toolkit";

interface InitialStateTypes {
    degree: string
}


const initialState: InitialStateTypes = {
    degree: "Celsius"
}

export const degreeSlice = createSlice({
    name: 'degreeSelection',
    initialState,
    reducers: {
        changeDegree: (state, action) => {
            state.degree = action.payload
        }
    }
})

export const { changeDegree }  = degreeSlice.actions;

export default degreeSlice.reducer;