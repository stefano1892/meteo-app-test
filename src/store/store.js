import { configureStore } from '@reduxjs/toolkit'
import weaterReducer from './weatherSlice'
import degreeReducer from './degreeSlice'

export const store = configureStore({
  reducer: {
    pastResearch: weaterReducer,
    degreeSelection: degreeReducer
  },
})