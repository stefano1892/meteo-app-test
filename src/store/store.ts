import { configureStore } from '@reduxjs/toolkit'
import weaterReducer from './weatherSlice'
import degreeReducer from './degreeSlice'

export const store = configureStore({
  reducer: {
    pastResearch: weaterReducer,
    degreeSelection: degreeReducer
  },
})

export type AppSelector = ReturnType<typeof store.getState>

// () => number => ReturnType<() => number> = number