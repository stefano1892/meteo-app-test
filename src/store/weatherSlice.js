import { createSlice } from "@reduxjs/toolkit";

const initialState = {
    searches: []
}

export const weatherSlice = createSlice({
    name: 'pastResearch',
    initialState,
    reducers: {
        addResearch: (state, action) => {
            state.searches.push(action.payload)
            if(state.searches.length > 5) {
                state.searches.shift()
            }
        }
    }
})

export const { addResearch }  = weatherSlice.actions;

export default weatherSlice.reducer;