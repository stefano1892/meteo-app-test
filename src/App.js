import './App.scss';
import 'bootstrap/dist/css/bootstrap.min.css';

import { Row, Col, Form, Button } from 'react-bootstrap'
import { useEffect, useState } from 'react';
import { getWeather } from './api/getCityWeather';
import { useDispatch } from 'react-redux';
import { addResearch } from './store/weatherSlice';
import { changeDegree } from './store/degreeSlice';
import { ErrorBoundary} from "react-error-boundary";

import FormComponent from './components/form-component';
import ActiveResultComponent from './components/active-result-component';
import PastResearchComponent from './components/past-research-component';
import NavbarComponent from './components/navbar-component';
import FooterComponent from './components/footer-component';

function App() {

  const dispatch = useDispatch()

  const [city, setCity] = useState('')
  const [cityResult , setCityResult] = useState()
  const [isError, setIsError] = useState(false) 

  useEffect(() => {
    const asyncFunction = async () => {
      try {
        const citySelected = await getWeather('rome');
        setCityResult(citySelected);
      } catch (error) {
        setIsError(true)
      }
    };
    asyncFunction();
  }, []);

  const ErrorFallback = () => {
    return (
      <div className='text-center mt-5'>
        <h2>Ops! La citt√† selezionata non esiste.</h2>
        <Button className='btn btn-primary' onClick={() => {
          setIsError(false)
          setCity('');
          window.location.reload()}}>Riprova</Button>
      </div>
    );
  }

  const handleInputChange = (e) => {
    setCity(e.target.value)
  }

  const handleSubmit = async (e) => {
    e.preventDefault()
    try {
      const citySelected = await getWeather(city);
      setCityResult(citySelected);
      dispatch(addResearch(citySelected));
      setCity('');
    } catch (error) {
      setIsError(true)
    }
  }

  const activateResearch = (city) => {
    setCityResult(city)
  }

  const handleChangeDegree = (e) => {
    dispatch(changeDegree(e.target.value))
  }

  return (
    <ErrorBoundary FallbackComponent={ErrorFallback} >
      <div className="App">
        <NavbarComponent />
        <Row className='m-0 w-100 mt-3'>
          <Col md={6} className='form-col'>
            <div className="d-flex">
              <FormComponent 
                handleInputChange = {handleInputChange}
                handleSubmit = {handleSubmit}
                city={city}
              />
            </div>
          </Col>
          <Col md={6}>
            <div className='degree-select-container'>
              <Form.Select className='degree-select' onChange={handleChangeDegree}>
                <option>Celsius</option>
                <option>Farenheit</option>
              </Form.Select>
            </div>
          </Col>
        </Row>
        {isError ? (
          <>
            <ErrorFallback />
          </>
        ) : (<>
          <ActiveResultComponent 
            city={cityResult}
          />
          <PastResearchComponent 
            activateResearch={activateResearch}
          />
        </>)}
        <footer>
          <FooterComponent/>
        </footer>
      </div>
    </ErrorBoundary>
  );
}

export default App;
