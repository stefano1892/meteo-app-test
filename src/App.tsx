import './App.scss';
import 'bootstrap/dist/css/bootstrap.min.css';

import { Row, Col, Form, Button } from 'react-bootstrap'
import React, { useCallback, useEffect, useState } from 'react';
import { getWeather } from './api/getCityWeather';
import { useDispatch } from 'react-redux';
import { addResearch } from './store/weatherSlice';
import { changeDegree } from './store/degreeSlice';
import { ErrorBoundary} from "react-error-boundary";

import FormComponent from './components/form-component';
import ActiveResultComponent from './components/active-result-component';
import PastResearchComponent from './components/past-research-component';
import NavbarComponent from './components/navbar-component';
import FooterComponent from './components/footer-component';
import { ICity } from './interfaces/ICity';

function App() {

  const dispatch = useDispatch()

  const [city, setCity] = useState<string>('')
  const [cityResult , setCityResult] = useState<ICity>()
  const [isError, setIsError] = useState<boolean>(false) 

  const getWeatherCallback = useCallback((city: string): Promise<ICity> => {
    return getWeather(city)
      .then(data => {
        setCityResult(data)
        setIsError(false)
        return data
      })
      .catch((error) => {
        setIsError(true)
        throw error
      })
  }, [])

  useEffect(() => {
    /*const asyncFunction = async (): Promise<void> => {
      try {
        const citySelected: ICity = await getWeather('rome');
        setCityResult(citySelected);
        setIsError(false)
      } catch (error) {
        setIsError(true)
      }
    };
    asyncFunction();*/
    getWeatherCallback('rome')
  }, []);

  const errorFallbackClick = useCallback(() => {
    setIsError(false)
    setCity('');
    window.location.reload()}, [])

  const ErrorFallback = () => {
    return (
      <div className='text-center mt-5'>
        <h2>Ops! La citt√† selezionata non esiste.</h2>
        <Button className='btn btn-primary' onClick={errorFallbackClick}>Riprova</Button>
      </div>
    );
  }

  const handleInputChange = (e: React.ChangeEvent<HTMLInputElement>): void => {
    setCity(e.target.value)
  }

  const handleSubmit = (e: React.FormEvent<HTMLFormElement>): void => {
    e.preventDefault()
    /*try {
      const citySelected: ICity = await getWeather(city);
      setCityResult(citySelected);
      dispatch(addResearch(citySelected));
      setCity('');
    } catch (error) {
      setIsError(true)
    }*/
    getWeatherCallback(city)
      .then(addResearch)
      .then(dispatch)
      .then(() => setCity(''))
    //   .then((citySelected) => {
    //     dispatch(addResearch(citySelected))
    //     setCity('')
    // })
  }

  const handleChangeDegree = (e: React.ChangeEvent<HTMLSelectElement>) => {
    dispatch(changeDegree(e.target.value))
  }

  return (
    <ErrorBoundary FallbackComponent={ErrorFallback} >
      <div className="App">
        <NavbarComponent />
        <Row className='m-0 w-100 mt-3'>
          <Col md={6} className='form-col'>
            <div className="d-flex">
              <FormComponent 
                handleInputChange = {handleInputChange}
                handleSubmit = {handleSubmit}
                city={city}
              />
            </div>
          </Col>
          <Col md={6}>
            <div className='degree-select-container'>
              <Form.Select className='degree-select' onChange={handleChangeDegree}>
                <option>Celsius</option>
                <option>Farenheit</option>
              </Form.Select>
            </div>
          </Col>
        </Row>
        {isError ? (
          <>
            <ErrorFallback />
          </>
        ) : (<>
          <ActiveResultComponent 
            city={cityResult}
          />
          <PastResearchComponent 
            activateResearch={setCityResult}
          />
        </>)}
        <footer>
          <FooterComponent/>
        </footer>
      </div>
    </ErrorBoundary>
  );
}

export default App;
