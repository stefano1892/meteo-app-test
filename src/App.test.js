// src/App.test.js
import React from 'react';
import { render, waitFor } from '@testing-library/react';
import App from './App';
import {store} from './store/store';
import { Provider } from 'react-redux';

describe('App', () => {
    test('input - exists', async () => {
        const { asFragment } = render(
            <Provider store={store}>
                <App />
            </Provider>
        );

        await waitFor(() => {
            expect(asFragment()).toMatchSnapshot();
        });
    });
});
