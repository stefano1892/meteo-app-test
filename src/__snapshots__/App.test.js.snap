// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`App input - exists 1`] = `
<DocumentFragment>
  <div
    class="App"
  >
    <nav
      class="navbar-light navbar navbar-expand-lg navbar-light"
    >
      <div
        class="container"
      >
        <a
          class="text-white navbar-brand"
          href="#home"
        >
          <b>
            BrandName
          </b>
        </a>
        <button
          aria-controls="responsive-navbar-nav"
          aria-label="Toggle navigation"
          class="navbar-toggler collapsed"
          type="button"
        >
          <span
            class="navbar-toggler-icon"
          />
        </button>
        <div
          class="navbar-collapse collapse"
          id="responsive-navbar-nav"
        >
          <div
            class="me-auto navbar-nav"
          >
            <a
              class="nav-link-custom active-link nav-link"
              data-rr-ui-event-key="#"
              href="#"
              role="button"
              tabindex="0"
            >
              Home
            </a>
            <a
              class="nav-link-custom nav-link"
              data-rr-ui-event-key="#"
              href="#"
              role="button"
              tabindex="0"
            >
              Previsioni
            </a>
            <a
              class="nav-link-custom nav-link"
              data-rr-ui-event-key="#"
              href="#"
              role="button"
              tabindex="0"
            >
              Situazione
            </a>
            <a
              class="nav-link-custom nav-link"
              data-rr-ui-event-key="#"
              href="#"
              role="button"
              tabindex="0"
            >
              Radar
            </a>
            <a
              class="nav-link-custom nav-link"
              data-rr-ui-event-key="#"
              href="#"
              role="button"
              tabindex="0"
            >
              Webcam
            </a>
            <a
              class="nav-link-custom nav-link"
              data-rr-ui-event-key="#"
              href="#"
              role="button"
              tabindex="0"
            >
              Video
            </a>
            <a
              class="nav-link-custom nav-link"
              data-rr-ui-event-key="#"
              href="#"
              role="button"
              tabindex="0"
            >
              News
            </a>
            <a
              class="nav-link-custom nav-link"
              data-rr-ui-event-key="#"
              href="#"
              role="button"
              tabindex="0"
            >
              Viabilità
            </a>
          </div>
          <div
            class="navbar-nav"
          >
            <a
              class="nav-link-custom nav-link"
              data-rr-ui-event-key="#"
              href="#"
              role="button"
              tabindex="0"
            >
              Business
            </a>
            <a
              class="nav-link-custom nav-link"
              data-rr-ui-event-key="2"
              href="#"
              role="button"
              tabindex="0"
            >
              Contatti
            </a>
          </div>
        </div>
      </div>
    </nav>
    <button
      data-testid="button-test"
    >
      cliccami
    </button>
    <div
      class="m-0 w-100 mt-3 row"
    >
      <div
        class="form-col col-md-6"
      >
        <div
          class="d-flex"
        >
          <form>
            <div
              class="form-container"
            >
              <input
                class="cityInput"
                placeholder="Inserisci la città..."
                type="text"
                value=""
              />
              <button
                class="btn btn-submit"
                disabled=""
                type="submit"
              >
                Cerca
              </button>
            </div>
          </form>
        </div>
      </div>
      <div
        class="col-md-6"
      >
        <div
          class="degree-select-container"
        >
          <select
            class="degree-select form-select"
          >
            <option>
              Celsius
            </option>
            <option>
              Farenheit
            </option>
          </select>
        </div>
      </div>
    </div>
    <div
      class="active-result-container"
    >
      <div
        class="m-0 w-100 row"
      >
        <div
          class="col-lg-6 col-md-6 col-sm-12 col-12"
        >
          <div
            class="main-container"
          />
        </div>
        <div
          class="col-lg-6 col-md-6 col-sm-12 col-12"
        >
          <div
            class="main-container responsive-item"
          />
        </div>
      </div>
    </div>
    <footer>
      <div
        class="w-100 mt-3 footer-containter"
      >
        <div
          class="w-100 m-0 row"
        >
          <div
            class="col"
          >
            <div
              class="footer-links-container text-center"
            >
              <b
                class="footer-link-title"
              >
                Link utili
              </b>
              <div
                class="mt-3"
              >
                <a
                  class="footer-link"
                  href="#"
                >
                  Home
                </a>
              </div>
              <div>
                <a
                  class="footer-link"
                  href="#"
                >
                  Previsioni
                </a>
              </div>
              <div>
                <a
                  class="footer-link"
                  href="#"
                >
                  Situazione
                </a>
              </div>
              <div />
            </div>
          </div>
          <div
            class="col"
          >
            <div
              class="footer-links-container text-center"
            >
              <b
                class="footer-link-title"
              >
                Novità
              </b>
              <div
                class="mt-3"
              >
                <a
                  class="footer-link"
                  href="#"
                >
                  Webcam
                </a>
              </div>
              <div>
                <a
                  class="footer-link"
                  href="#"
                >
                  News
                </a>
              </div>
              <div>
                <a
                  class="footer-link"
                  href="#"
                >
                  Viabilità
                </a>
              </div>
            </div>
          </div>
          <div
            class="col"
          >
            <div
              class="footer-links-container text-center"
            >
              <b
                class="footer-link-title"
              >
                Contatti
              </b>
              <div
                class="mt-3"
              >
                <a
                  class="footer-link"
                  href="#"
                >
                  Contatti
                </a>
              </div>
              <div>
                <a
                  class="footer-link"
                  href="#"
                >
                  Business
                </a>
              </div>
              <div>
                <a
                  class="footer-link"
                  href="#"
                >
                  Lavora con noi
                </a>
              </div>
            </div>
          </div>
        </div>
        <div
          class="registered-mark"
        >
          © 2024 Copyright: Brandname.com
        </div>
      </div>
    </footer>
  </div>
</DocumentFragment>
`;
